driverdataaggregation/main.py

```diff
@@ -0,0 +1,601 @@
+        :param run_db_as_batch: run database as separate batch submission
+                                each iteration
+        :type run_db_as_batch: bool, optional
+        :param batch_args: additional batch args for the database
+        :type batch_args: dict, optional
+        :param db_hosts: optionally supply hosts to launch the database on
+        :type db_hosts: list, optional
+        :param db_nodes: number of compute hosts to use for the database
+        :type db_nodes: list, optional
+        :param db_cpus: number of cpus per compute host for the database
+        :type db_cpus: list, optional
+        :param db_port: port to use for the database
+        :type db_port: int, optional
+        :param net_ifname: network interface to use i.e. "ib0" for infiniband or
+                           "ipogif0" aries networks
+        :type net_ifname: str, optional
+        :param clients_per_node: client tasks per compute node for the aggregation
+                                 producer app
+        :type clients_per_node: list, optional
+        :param client_nodes: number of compute nodes to use for the aggregation
+                             producer app
+        :type client_nodes: list, optional
+        :param iterations: number of append/retrieve loops run by the applications
+        :type iterations: int
+        :param tensor_bytes: list of tensor sizes in bytes
+        :type tensor_bytes: list[int], optional
+        :param tensors_per_dataset: list of number of tensors per dataset
+        :type tensor_bytes: list[int], optional
+        :param client_threads: list of the number of client threads used for data
+                               aggregation
+        :type client_threads: list[int], optional
+        :param cpu_hyperthreads: the number of hyperthreads per cpu.  This is done
+                                 to request that the consumer application utilizes
+                                 all physical cores for each client thread.
+        :type cpu_hyperthreads: int, optional
+        """
+        logger.info("Starting dataset aggregation scaling tests")
+        logger.info(f"Running with database backend: {_get_db_backend()}")
+        logger.info(f"Running with launcher: {launcher}")
+
+        exp = create_folder(exp_name, launcher)
+
+        for db_node_count in db_nodes:
+
+            # start the database only once per value in db_nodes so all permutations
+            # are executed with the same database size without bringin down the database
+            db = start_database(exp,
+                                db_port,
+                                db_node_count,
+                                db_cpus,
+                                None, # not setting threads per queue in throughput tests
+                                net_ifname,
+                                run_db_as_batch,
+                                batch_args,
+                                db_hosts)
```

Comment:
I believe you were getting this error bc I had it to where u needed to pass in a constraint val, now ive added a default so it should take care of this issue?

Reply:
Related, but subtly different:

The old fn signature of `start_database` (with type hints added for clarity)  was:
```py
# old
def start_database(
    exp: Experiment,
    port: int,
    nodes: int,
    cpus: int,
    tpq: int | None,
    net_ifname: str,
    run_as_batch: bool,
    batch_args: dict[str, str],
    hosts: list[str]
) -> Orchestrator:
    ...
```
and the new signature (again types added for clarity) is
```py
# new
def start_database(
    exp: Experiment
    node_feature: str,  # <-- Note: the type of this 2nd arg changed from `int` -> `str`
    port: int,
    nodes: int,
    cpus: int,  # Note again the type of this 5th arg changed from `int | None` -> `int`
    tpq: int | None,  # <-- and the type of the 6th arg changed from `str` -> `int | None`
    net_ifname: str,  # <-- and the 7th from `bool` -> `str`
    run_as_batch: bool,   # <-- As did the 8th from `dict[str, str]` -> `bool`
    hosts: list[str]
) -> Orchestrator:
    ...
```
Currently, at this line, the 5th arg being passed is `None`, but `start_database` just takes that param and passess it to `Experiment.create_database(..., cpu=..., ...)`, which and the `cpu` param must be of type `int`. So because python does not allow casting `None` it an `int` it raises a type error.

Reply:
Long story short is that it looks like the fn definition of `start_database` was modified, but the call signatures in the throughput/aggregation studies were not updated to reflect the change